pkg/reconciler/extension/extension.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2022 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package extension
+
+import (
+	"context"
+
+	autoscalingv1alpha1 "knative.dev/serving/pkg/apis/autoscaling/v1alpha1"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// ServingExtension enables pluggable extensive features added into the controller
+type ServingExtension interface {
+	// TransformRevision transforms the existing revision.
+	TransformRevision(*v1.Revision) *v1.Revision
```

Comment:
What transformations are you expecting to do on the revision?

Reply:
The only change is to fill in the transformed revision with the `minScale and maxScale information` for each stage, and leverage them to create or update the podautoscaler, so that podautoscaler picks up the information about how many pods to launch for each stage.

Reply:
Could you add this back to the document or is it mentioned somewhere in the implementation?
@dprotaso @houshengbo the more I read this I feel that the `ServiceOrchestrator` could be also something added as a background routine within Serving repo behind some flag and so we dont need these kind of extensions. More like the multi-scaler plus some internal data structure like the Decider.
Eventually I think this is a performance bug we are trying to fix and when this get's mature, it's use should be transparent.